### PR TITLE
Add QueuingContext for queue interactions

### PR DESCRIPTION
**Pull request by johannesjmeyer**
**Context:**
A refactor of the queuing system was proposed in the corresponding ADR to replace the poorly encapsulated queue interactions relying on `qml._current_context`.

**Description of the Change:**
An abstract base class `QueuingContext` oversees all queue interactions. Calls to `_current_context._append_op` and `_current_context._remove_op` where replaced with `QueuingContext.append_operation` and `.remove_operation` calls.

Due to this, the code for the `OperationRecorder` class simplified a lot and the underlying `Recorder` (basically a mock QNode) class is no longer necessary.

**Benefits:**
A clearer and more versatile QueuingInterface

### CODE REVIEW

**Comment by johannesjmeyer** [comment]
I also tried to tackle issue #360 in this PR by removing the deletion and reinsertion of the observable in the measurement routines. But this caused the check that all observables have to be returned in order of creation to fail. But this will also be updated (and measure.py greatly simplified) when the other parts of the queuing ADR are being implemented.

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -93,15 +93,21 @@
...
-  ...     qml.S(wires=1).inv
+  ...     qml.S(wires=1).inv()
   ...     return qml.probs(wires=[0, 1])
   >>> circuit(0.2)
   array([0.99003329, 0.        , 0.        , 0.00996671])
   >>> print(circuit.draw())
-  0: ──RX(0.2)──╭C─────╭┤ Probs
-  1: ───────────╰X──S──╰┤ Probs
+  0: ──RX(0.2)──╭C───────╭┤ Probs
+  1: ───────────╰X──S⁻¹──╰┤ Probs
```
Comment (josh146) [inline .github/CHANGELOG.md]: Oops, good catch!

File: pennylane/__init__.py [hunk 2]
```diff
@@ -23,6 +23,8 @@
 
 from semantic_version import Version, Spec
 
+# QueuingContext needs to be imported before all other pennylane imports
+from ._queuing_context import QueuingContext  # pylint: disable=wrong-import-order
```
Comment (josh146) [inline pennylane/__init__.py]: :+1:

File: pennylane/_queuing_context.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+"""
+import abc
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for Operations.
+    """
+
+    _active_contexts = []
+    """The list of contexts that are currently active."""
```
Comment (josh146) [inline pennylane/_queuing_context.py]: This is a big change right? Now, multiple contexts can be available at any one time? Probably doesn't make a difference except for when parallel/threading is used

File: pennylane/_queuing_context.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            QueuingContext: This instance
+        """
+        QueuingContext._active_contexts.append(self)
+
+        return self
+
+    def __exit__(self, exception_type, exception_value, traceback):
+        """Remove this instance from the global list of active contexts.
+        """
+        QueuingContext._active_contexts.remove(self)
```
Comment (josh146) [inline pennylane/_queuing_context.py]: Is it possible for the same QNode to be added to the context list twice? In which case, will this remove method potentially remove the wrong QNode?

File: pennylane/_queuing_context.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    @classmethod
+    def append_operator(cls, operator):
+        """Append an operator to the global queue(s).
+
+        Args:
+            operator (Operator): The Operator instance to be appended
+        """
+        for context in cls._active_contexts:
+            context._append_operator(operator)  # pylint: disable=protected-access
```
Comment (josh146) [inline pennylane/_queuing_context.py]: Just clarifying, but this adds the operation to _all_ contexts in the context list?

Comment (josh146) [reply pennylane/_queuing_context.py]: What is the use case you are envisioning here?

File: pennylane/measure.py [hunk 6]
```diff
@@ -18,14 +18,8 @@
...
 """
 import pennylane as qml
-from .operation import Observable, Sample, Variance, Expectation, Probability, Tensor
-from .qnodes import QuantumFunctionError
-
-
-def _remove_if_in_queue(op):
-    r"""Helper function to handle removing ops from the QNode queue"""
-    if op in qml._current_context.queue:
-        qml._current_context.queue.remove(op)
```
Comment (josh146) [inline pennylane/measure.py]: Nice to finally be able to remove this!

File: pennylane/measure.py [hunk 7]
```diff
@@ -60,20 +54,15 @@ def circuit(x):
...
-        if isinstance(op, Tensor):
-            for o in op.obs:
-                _remove_if_in_queue(o)
-        else:
-            _remove_if_in_queue(op)
+    if isinstance(op, Tensor):
+        for o in op.obs:
+            qml.QueuingContext.remove_operator(o)
+    else:
+        qml.QueuingContext.remove_operator(op)
```
Comment (josh146) [inline pennylane/measure.py]: I feel like this if statement could also be cleaned up, by having the Tensor class encapsulate queuing logic. But this is likely a separate PR.

File: pennylane/operation.py [hunk 8]
```diff
@@ -232,10 +232,7 @@ class Operator(abc.ABC):
         wires (Sequence[int]): Subsystems it acts on. If not given, args[-1]
             is interpreted as wires.
         do_queue (bool): Indicates whether the operator should be
```
Comment (josh146) [inline pennylane/operation.py]: Is `do_queue` needed any more? ever since we moved to having it False by default, I think it has been redundant. (Its original purpose was because queuing used to be _true_ by default, so we needed a way to turn it off).

File: pennylane/qnodes/base.py [hunk 9]
```diff
@@ -314,48 +298,39 @@ def _op_descendants(self, op, only):
...
+                self.queue.append(operator)
             else:
-                self.obs_queue.append(op)
+                self.obs_queue.append(operator)
         else:
             if self.obs_queue:
                 raise QuantumFunctionError(
                     "State preparations and gates must precede measured observables."
                 )
-            self.queue.append(op)  # TODO rename self.queue to self.op_queue
```
Comment (josh146) [inline pennylane/qnodes/base.py]: Might as well leave the TODO in?

File: pennylane/utils.py [hunk 10]
```diff
@@ -291,32 +242,21 @@ class OperationRecorder:
     """
 
     def __init__(self):
-        self.rec = None
-
-        self.queue = None
+        self.queue = []
```
Comment (josh146) [inline pennylane/utils.py]: So much cleaner 😍

File: tests/qnodes/test_qnode_base.py [hunk 11]
```diff
@@ -348,24 +348,6 @@ def circuit(x):
 class TestQNodeExceptions:
     """Tests that QNode raises proper errors"""
 
-    def test_current_context_modified_outside_construct(
-        self, operable_mock_device_2_wires, monkeypatch
-    ):
-        """Error: _current_context was modified outside of construct."""
```
Comment (josh146) [inline tests/qnodes/test_qnode_base.py]: Is this edge case no longer possible?

File: pennylane/_queuing_context.py (hunk 3 above)
Comment (johannesjmeyer) [reply pennylane/_queuing_context.py]: Yes, but this is basically inteded as 1 qnode + X operation recorders. 

Parallel construction of QNodes isn't possible with PL's Operation queue structure anyways, so this is not deteriorating at this end either.

Comment (josh146) [reply pennylane/_queuing_context.py]: Ah, that makes sense

File: pennylane/_queuing_context.py (hunk 4 above)
Comment (johannesjmeyer) [reply pennylane/_queuing_context.py]: Only in the core code, because we have the QNode use the context manager template
```with self:
        ... all operations created here go into the qnode ...
```
For a second QNode to go active, one would have to _construct a qnode inside the with block of the qnode (which is also in _construct).

And for the same QNode, this is basically not possible currently.

File: pennylane/_queuing_context.py (hunk 5 above)
Comment (johannesjmeyer) [reply pennylane/_queuing_context.py]: Exactly! Active contexts are usually 0-1 qnodes and 0-X operation recorders, and I want the operation go to all of them.

File: pennylane/measure.py (hunk 6 above)
Comment (johannesjmeyer) [reply pennylane/measure.py]: "remove if in queue" is now the default behaviour of `QueuingContext.remove_operation`.

File: pennylane/measure.py (hunk 7 above)
Comment (johannesjmeyer) [reply pennylane/measure.py]: This is exactly what I proposed in parts B and C of the queuing ADR

File: pennylane/operation.py (hunk 8 above)
Comment (johannesjmeyer) [reply pennylane/operation.py]: I guess also that it is redundant now. The only thing one could do is to actively block queuing when inside a QNode for example, but I don't see the use case

File: pennylane/qnodes/base.py (hunk 9 above)
Comment (johannesjmeyer) [reply pennylane/qnodes/base.py]: Yes, I think I'll have to do a renaming PR soon!

File: tests/qnodes/test_qnode_base.py (hunk 11 above)
Comment (johannesjmeyer) [reply tests/qnodes/test_qnode_base.py]: I just assumed that nobody messes with `QueuingContext._active_contexts`. If they do it's their fault.

**Comment by josh146** [comment]
Thanks @johannesjmeyer, that answers most of my questions. Happy for this to be merged in after @antalszava's had a chance to look it over.

File: pennylane/__init__.py (hunk 2 above)
Comment (antalszava) [reply pennylane/__init__.py]: Out of curiosity: how come the order is set here like this?

File: pennylane/_queuing_context.py [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    @classmethod
+    def remove_operator(cls, operator):
+        """Remove an operator from the global queue(s) if it is in the queue(s).
+
+        Args:
+            operator (Operator): The Operator instance to be removed
+        """
+        for context in cls._active_contexts:
+            # We use the duck-typing approach to assume that the underlying remove
+            # behaves like list.remove and throws an ValueError if the operator
```
Comment (antalszava) [inline pennylane/_queuing_context.py]:
```suggestion
            # behaves like list.remove and throws a ValueError if the operator
```

File: pennylane/_queuing_context.py (hunk 5 above)
Comment (antalszava) [reply pennylane/_queuing_context.py]: Perhaps this could be included in the docstring to make it clearer :)

File: pennylane/measure.py [hunk 13]
```diff
@@ -18,14 +18,8 @@
...
-from .operation import Observable, Sample, Variance, Expectation, Probability, Tensor
-from .qnodes import QuantumFunctionError
-
-
-def _remove_if_in_queue(op):
-    r"""Helper function to handle removing ops from the QNode queue"""
-    if op in qml._current_context.queue:
-        qml._current_context.queue.remove(op)
+from pennylane.operation import Observable, Sample, Variance, Expectation, Probability, Tensor
+from pennylane.qnodes import QuantumFunctionError
```
Comment (antalszava) [inline pennylane/measure.py]: Was there a need to switch to absolute imports or was it more so just a minor design decision? (Looks good from my side, just curious :) )

File: pennylane/_queuing_context.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`QueuingContext` abstract base class.
+"""
+import abc
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for Operations.
```
Comment (antalszava) [inline pennylane/_queuing_context.py]: An extra sentence with an example could be helpful here. This one that was previously used for the `Recorder` is pretty good in my opinion:
`"The Recorder class is a very minimal QNode, that simply provides a QNode context for operator queueing."`

Overall, from my side something along the lines of answering how defining a `QueuingContext` comes in to play (e.g. for `QNodes`, operation recorders, etc.) would be perfect!

File: pennylane/measure.py [hunk 15]
```diff
@@ -60,20 +54,15 @@ def circuit(x):
...
+    else:
+        qml.QueuingContext.remove_operator(op)
 
-    # set return type to be an expectation value
     op.return_type = Expectation
 
-    if qml._current_context is not None:
-        # add observable to QNode observable queue
-        qml._current_context._append_op(op)
+    qml.QueuingContext.append_operator(op)
```
Comment (antalszava) [inline pennylane/measure.py]: These look indeed really clean! :)

File: pennylane/__init__.py (hunk 2 above)
Comment (johannesjmeyer) [reply pennylane/__init__.py]: `QueuingContext` now acts as the baseclass for `BaseQNode` and `OperationRecorder`, therefore I have to import it before those classes are created, i.e. in the subsequent imports.

File: tests/test_utils.py [hunk 16]
```diff
@@ -303,126 +303,15 @@ def test_expand_three_nonconsecutive_nonascending_wires(self, tol):
         assert np.allclose(res, expected, atol=tol, rtol=0)
 
 
-class TestRecorder:
```
Comment (antalszava) [inline tests/test_utils.py]: Minor suggestion here (not too crucial): small unit tests for the behaviour of the `QueuingContext` could be included. An example could be creating a test class, that inherits from `QueuingContext` and checking that each of the standard functionalities in there indeed works the way they should (e. g. `remove_operator` indeed throwing an error in certain cases).

Overall this would more so be a sanity check and an indicator that there is something going wrong there, should parts of the logic change.

File: pennylane/measure.py (hunk 13 above)
Comment (johannesjmeyer) [reply pennylane/measure.py]: It made Codecov stop complaining about circular imports 😄

File: pennylane/_queuing_context.py (hunk 14 above)
Comment (johannesjmeyer) [reply pennylane/_queuing_context.py]: Please see my take on it! Feel free to edit it tough, I'm not 100% happy.

File: tests/test_utils.py (hunk 16 above)
Comment (johannesjmeyer) [reply tests/test_utils.py]: Yes, that is totally right. I'll add some tests!

File: pennylane/_queuing_context.py [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+    behaviour. As the constructor does not know to which QNode (PennyLane's circuit
+    object) it should attach the quantum gate we need to use a global variable that
+    holds that information. The ``QueuingContext`` class realizes this by providing
+    access to the current QNode. Furthermore, it provides the flexibility to have
+    multiple objects record the creation of quantum gates.
+
+    The QueuingContext class both acts as the abstract base class for all
+    classes that expose a queue for Operations (so-called contexts), as well
+    as the interface to said queues. The active contexts contain maximally one QNode
+    and an arbitrary number of other contexts like the OperationRecorder.
```
Comment (antalszava) [inline pennylane/_queuing_context.py]: So great to have this point here, thanks for adding it :)

File: pennylane/_queuing_context.py [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for Operations.
+
+    In PennyLane, quantum gates are not attached to some kind of circuit object
+    by the user. Here, the *construction* of a quantum gate is used to realize this
+    behaviour. As the constructor does not know to which QNode (PennyLane's circuit
+    object) it should attach the quantum gate we need to use a global variable that
+    holds that information. The ``QueuingContext`` class realizes this by providing
+    access to the current QNode. Furthermore, it provides the flexibility to have
+    multiple objects record the creation of quantum gates.
```
Comment (antalszava) [inline pennylane/_queuing_context.py]: This part is a great description of the problem that is being solved here! One thing I'm wondering if this would be the place where it should go in. Maybe it would be cool if we created a `Architecture` page or something similar in the docs, and this would be a perfect section there (then there could be a link included here).

Overall this would for sure be out of scope for this PR I believe, so for now maybe we could just keep this description here.

Comment (johannesjmeyer) [reply pennylane/_queuing_context.py]: Yes, but I don't know what else would go into that `Architecture` page 🤔 But that's surely for another PR

File: pennylane/_queuing_context.py [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+"""
+import abc
+
+
+class QueuingContext(abc.ABC):
+    """Abstract base class for classes that exposes a queue for Operations.
+
+    In PennyLane, quantum gates are not attached to some kind of circuit object
+    by the user. Here, the *construction* of a quantum gate is used to realize this
+    behaviour. As the constructor does not know to which QNode (PennyLane's circuit
```
Comment (antalszava) [inline pennylane/_queuing_context.py]: A suggestion here (subjective, feel free to discard it): I would suggest rewording the first sentence as follows (not too sure about the second sentence and later the third might have to be rephrased since the `QNode` was introduced there).
```suggestion
    In PennyLane, the construction of quantum gates is separated from the specific
    quantum node (:class:`BaseQNode`) that they belong to. As the
    constructor does not know to which QNode (PennyLane's circuit
```

Comment (johannesjmeyer) [reply pennylane/_queuing_context.py]: Sorry, I already merged because you approved and didn't write anything further -- I like that wording better though.
